src/background.js

```diff
@@ -173,36 +173,30 @@ const messageOps = {
 function syncBookmarks(items) {
   const title = browser.i18n.getMessage('bookmarkFolderTitle');
   browser.bookmarks.search({title: title}).then(folders => {
-    return Promise.all(folders.map(folder => {
-      return browser.bookmarks.removeTree(folder.id);
-    }));
-  }).then(() => {
+    if (folders.length) {
+      return folders[0];
+    }
     return browser.bookmarks.create({title: title});
   }).then(snoozeTabsFolder => {
     log('Sync Folder!', snoozeTabsFolder, Object.values(items));
-    const sortedEntries = [...Object.values(items)];
-    sortedEntries.sort((a, b) => {
-      if (a.time === NEXT_OPEN) {
-        if (b.time === NEXT_OPEN) {
-          return -(a.title.localeCompare(b.title));
-        } else {
-          return 1;
-        }
-      } else if (b.time === NEXT_OPEN) {
-        return -1;
-      } else {
-        return b.time - a.time;
-      }
-    });
+    return browser.bookmarks.getChildren(snoozeTabsFolder.id).then((bookmarks) => {
+      const tabs = [...Object.values(items)];
+      const toCreate = tabs.filter((tab) => !bookmarks.find((bookmark) => tab.url === bookmark.url));
+      const toRemove = bookmarks.filter((bookmark) => !tabs.find((tab) => bookmark.url === tab.url));
 
-    return Promise.all(sortedEntries.map(item => {
-      return browser.bookmarks.create({
-        parentId: snoozeTabsFolder.id,
-        title: item.title,
-        url: item.url,
-        index: 0
-      });
-    }));
+      const operations = toCreate.map(item => {
+        log(`Creating ${item.url}.`);
+        return browser.bookmarks.create({
```

Comment:
Unless `browser.bookmarks` is smart — and looking at what I think is the file that implements it in `m-c`, it's not — this is a performance nightmare: `N` SQL reads and writes (however many are needed for a create or a delete) for each change, outside a transaction. That's a lot of filesystem operations and a lot of hits on Places.

Further, a bookmarks sync will be triggered by the _first_ write, not the last. At best, with all of Kit and co's work, each time you flush this to disk with more than one change you'll get _two_ syncs.

`browser.bookmarks` considered very harmful: it's simply a bad API. There's not much _you_ can do about it, other than to not use it.